src/compartment.c

```diff
@@ -1288,16 +1127,79 @@ resolve_comp_tls_regions(struct Compartment *new_comp)
     new_comp->libs_tls_sects->region_size = comp_tls_size;
 }
 
+static void
+stage_comp(struct Compartment *to_stage)
+{
+    void *base_stage_addr = mmap(NULL, to_stage->total_size,
+        PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+    if (base_stage_addr == MAP_FAILED)
+    {
+        err(1, "Error staging compartment %zu", to_stage->id);
+    }
+
+    // Copy over segment data and `.tdata`
+    struct LibDependency *lib_dep;
+    struct SegmentMap lib_dep_seg;
+    size_t tls_allocd = 0x0;
+    for (size_t i = 0; i < to_stage->libs_count; ++i)
+    {
+        lib_dep = to_stage->libs[i];
+
+        // Segment data
+        for (size_t j = 0; j < lib_dep->lib_segs_count; ++j)
+        {
+            lib_dep_seg = lib_dep->lib_segs[j];
+            memcpy(get_seg_target(base_stage_addr, lib_dep, j),
+                (char *) lib_dep->data_base + lib_dep_seg.offset,
+                lib_dep_seg.file_sz);
+        }
+
+        // `.tdata` for TLS
+        if (to_stage->libs[i]->tls_sec_size != 0)
+        {
+            assert(to_stage->libs[i]->tls_sec_addr != 0x0);
+            memcpy((char *) base_stage_addr
+                    + (uintptr_t) to_stage->libs_tls_sects->region_start
+                    + (uintptr_t) tls_allocd,
+                (char *) lib_dep->data_base
+                    + (uintptr_t) to_stage->libs[i]->tls_sec_off,
+                lib_dep->tls_sec_size);
+            tls_allocd += to_stage->libs[i]->tls_sec_size;
+        }
+
+        munmap(lib_dep->data_base, lib_dep->data_size);
+        lib_dep->data_base = NULL;
+        lib_dep->data_size = 0;
+    }
+
+    /* Copy over environ variables
+     *
+     * We need a pointer to an array of string pointers, so we synthetically
+     * create one. We don't expect this pointer to move, as the maximum allowed
+     * size for the `environ` array is already allocated
+     */
+
+    void *environ_addr
+        = (char *) to_stage->environ_ptr + (uintptr_t) base_stage_addr;
+    *((uintptr_t *) environ_addr) = sizeof(void *);
+    environ_addr = (char *) environ_addr + sizeof(char *);
+
+    // Copy over prepared `environ` data from manager
+    memcpy(environ_addr, to_stage->cc->env_ptr, to_stage->cc->env_ptr_sz);
```

Comment:
Bitter experience has taught me to prefer `memmove` just because `memcpy` *requires* that the two buffers don't overlap. I think in this case that's correct, but I've seen more than one bug when someone's got that (generally performance-irrelevant) invariant wrong. Up to you though!

Reply:
These actually *must* be non-overlapping, and they should be so due to the nature of the two `malloc`s used to create the two memory regions - [1] for `env_ptr`, and [2] for the region containing `environ_addr`. I maybe could add a check to ensure that the staging area for the compartment doesn't overlap the manager's working memory? Alternatively, I'd leave as is, and perhaps add the check in a test, or leave future bugs to be found by using `asan` / `msan`.

[1] https://github.com/capablevms/CHERI-ELF-comp/blob/master/src/manager.c#L369
[2] https://github.com/capablevms/CHERI-ELF-comp/pull/42/files#diff-7887211152112561f3aeba0bbee694021f8d663651ae72fe0164649b2ba2ea32R1133